src/server/chunk_server/chunk_server_impl.cc

```diff
@@ -0,0 +1,129 @@
+  chunk_server_clients_.clear();
+  chunk_versions_.clear();
+  lease_and_expiration_unix_sec_.clear();
+}
+
+StatusOr<ChunkServerImpl*> ChunkServerImpl::ConstructChunkServerImpl(
+    const std::string& config_filename, const std::string& chunk_server_name,
+    const bool resolve_hostname) {
+  LOG(INFO) << "Parsing configuration file...";
+  // Instantiate a ConfigManager with the given filename
+  StatusOr<ConfigManager*> config_manager_or(
+      ConfigManager::GetConfig(config_filename));
+  if (!config_manager_or.ok()) {
+    return config_manager_or.status();
+  }
+  return new ChunkServerImpl(config_manager_or.ValueOrDie(), chunk_server_name,
+                             resolve_hostname);
+}
+
+//
+// Lease Management
+//
+
+void ChunkServerImpl::AddOrUpdateLease(const std::string& file_handle,
+                                       const uint64_t expiration_unix_sec) {
+  lease_and_expiration_unix_sec_[file_handle] = expiration_unix_sec;
+}
+
+bool ChunkServerImpl::HasWriteLease(const std::string& file_handle) {
+  if (!lease_and_expiration_unix_sec_.contains(file_handle)) {
+    return false;
+  }
+  // there could be a time-of-check and time-of-use race condition, but we will
+  // allow it happen for now, and simply let caller retry
+  absl::Time expiration_time =
+      absl::FromUnixSeconds(lease_and_expiration_unix_sec_[file_handle]);
+  return absl::Now() < expiration_time;
+}
+
+StatusOr<absl::Time> ChunkServerImpl::GetLeaseExpirationTime(
+    const std::string& file_handle) {
+  if (!lease_and_expiration_unix_sec_.contains(file_handle)) {
+    return Status(
+        google::protobuf::util::error::NOT_FOUND,
+        absl::StrCat("Lease is not found for file handle: ", file_handle));
+  } else {
+    return absl::FromUnixSeconds(lease_and_expiration_unix_sec_[file_handle]);
+  }
+}
+
+void ChunkServerImpl::RemoveLease(const std::string& file_handle) {
+  lease_and_expiration_unix_sec_.erase(file_handle);
+}
+
+//
+// Mock Functions - File Chunks management
+//
+
+void ChunkServerImpl::SetChunkVersion(const std::string& file_handle,
+                                      const uint32_t version) {
+  // TODO(tugan,michael): use chunk file manager instead, when ready
+  chunk_versions_[file_handle] = version;
+}
+
+google::protobuf::util::StatusOr<uint32_t> ChunkServerImpl::GetChunkVersion(
+    const std::string& file_handle) {
+  // TODO(tugan,michael): use chunk file manager instead, when ready
+  if (!chunk_versions_.contains(file_handle)) {
+    return Status(google::protobuf::util::error::NOT_FOUND,
+                  absl::StrCat("File chunk is not found: ", file_handle));
+  } else {
+    return chunk_versions_[file_handle];
+  }
+}
+
+//
+// gRPC Protocol Helpers
+//
+
+std::shared_ptr<MasterChunkServerManagerServiceClient>
+ChunkServerImpl::GetMasterProtocolClient(const std::string& server_address) {
+  if (master_server_clients_.contains(server_address)) {
+    return master_server_clients_[server_address];
```

Comment:
I presume that there is no deletion happening to this collection so this is safe? Since master_server_clients_ is a thread-safe hash map this line could be subject to time-of-check and time-of-use issue. I had introduced common::parallel_hash_map with some revised methods to take care issues like this. Probably I can make that change if necessary later, we don't have to be blocked by this right now.